lib/kube/proxy/streamproto/proto.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package streamproto
+
+import (
+	"io"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/client-go/tools/remotecommand"
+)
+
+// metaMessage is a control message containing one or more payloads.
+type metaMessage struct {
+	Resize          *remotecommand.TerminalSize `json:"resize,omitempty"`
+	ForceTerminate  bool                        `json:"force_terminate,omitempty"`
+	ClientHandshake *ClientHandshake            `json:"client_handshake,omitempty"`
+	ServerHandshake *ServerHandshake            `json:"server_handshake,omitempty"`
+}
+
+// ClientHandshake is the first message sent by a client to inform a server of it's intentions.
+type ClientHandshake struct {
+	Mode types.SessionParticipantMode `json:"mode"`
+}
+
+// ServerHandshake is the first message sent by a server to inform a client of the session settings.
+type ServerHandshake struct {
+	MFARequired bool `json:"mfa_required"`
+}
+
+// SessionStream represents one end of the bidirectional session connection.
+type SessionStream struct {
+	// The underlying websocket connection.
+	conn *websocket.Conn
+
+	// A stream of incoming session packets.
+	in chan []byte
+
+	// Optionally contains a partially read session packet.
+	currentIn []byte
+
+	// A list of resize requests.
+	resizeQueue chan *remotecommand.TerminalSize
+
+	// A notification channel for force termination requests.
+	forceTerminate chan struct{}
+
+	writeSync   sync.Mutex
+	done        chan struct{}
+	closedC     sync.Once
```

Comment:
```suggestion
	once     sync.Once
```

I would call this `once` or `closeOnce`. `closedC` sounds like a channel to me.